unit/analyses/constant_propagator.cpp

```diff
@@ -113,4 +120,271 @@ SCENARIO("constant_propagator", "[core][analyses][constant_propagator]")
+  {
+    // Create a program like:
+    // bool b;
+    // if(!b)
+    //   b = true;
+    // Repeat this using bool_typet and c_bool_typet for "bool".
+
+    stream_message_handlert log(std::cout);
+
+    goto_modelt goto_model;
+    namespacet ns(goto_model.symbol_table);
+
+    symbolt bool_local;
+    symbolt c_bool_local;
+    bool_local.name = "bool_local";
+    bool_local.type = bool_typet();
+    bool_local.mode = ID_C;
+    c_bool_local.name = "c_bool_local";
+    c_bool_local.type = c_bool_typet(8);
+    c_bool_local.mode = ID_C;
+
+    code_blockt code;
+    code.copy_to_operands(code_declt(bool_local.symbol_expr()));
+    code.copy_to_operands(code_declt(c_bool_local.symbol_expr()));
+
+    code_ifthenelset bool_cond_block;
+    bool_cond_block.cond() = not_exprt(bool_local.symbol_expr());
+    bool_cond_block.then_case() =
+      code_assignt(bool_local.symbol_expr(), true_exprt());
+
+    code_ifthenelset c_bool_cond_block;
+    const exprt c_bool_true = from_integer(1, c_bool_typet(8));
+    c_bool_cond_block.cond() =
+      notequal_exprt(c_bool_local.symbol_expr(), c_bool_true);
+    c_bool_cond_block.then_case() =
+      code_assignt(c_bool_local.symbol_expr(), c_bool_true);
+
+    code.move_to_operands(bool_cond_block);
+    code.move_to_operands(c_bool_cond_block);
+
+    symbolt main_function_symbol;
+    main_function_symbol.name = "main";
+    main_function_symbol.type = code_typet();
+    main_function_symbol.value = code;
+    main_function_symbol.mode = ID_C;
+
+    goto_model.symbol_table.add(bool_local);
+    goto_model.symbol_table.add(c_bool_local);
+    goto_model.symbol_table.add(main_function_symbol);
+
+    goto_convert(goto_model, log);
+
+    const goto_functiont &main_function = goto_model.get_goto_function("main");
+
+    // Find the first DEAD instruction -- we will test our results there, after
+    // the function body but before the exit sequence.
+    goto_programt::const_targett test_instruction =
+      main_function.body.instructions.begin();
+    while(
+      test_instruction != main_function.body.instructions.end() &&
+      !test_instruction->is_dead())
+    {
+      ++test_instruction;
+    }
+
+    REQUIRE(test_instruction != main_function.body.instructions.end());
+
+    WHEN("Constant propagation is performed")
+    {
+      constant_propagator_ait constant_propagator(main_function);
+      constant_propagator(main_function, ns);
+
+      THEN("The propagator should conclude that both booleans are true at the "
+           "end of the function")
+      {
+        const auto &final_domain = constant_propagator[test_instruction];
+
+        REQUIRE(final_domain.values.is_constant(bool_local.symbol_expr()));
+        REQUIRE(final_domain.values.is_constant(c_bool_local.symbol_expr()));
+      }
+    }
+  }
+
+  GIVEN("A GOTO program testing ways of expressing boolean tests")
+  {
+    // Create a program like:
+    // bool b0, b1, b2, ...;
+    // int marker;
+    // if(b0)
```

Comment:
Would it be possible / reasonable to break this up into multiple smaller examples? Something this big is a bit difficult to follow.